modules/store/src/reducer_creator.ts

```diff
@@ -11,177 +11,25 @@ export interface OnReducer<S, C extends ActionCreator[]> {
-): On<S>;
-export function on<
-  C1 extends ActionCreator,
-  C2 extends ActionCreator,
-  C3 extends ActionCreator,
-  C4 extends ActionCreator,
-  C5 extends ActionCreator,
-  C6 extends ActionCreator,
-  C7 extends ActionCreator,
-  S
->(
-  creator1: C1,
-  creator2: C2,
-  creator3: C3,
-  creator4: C4,
-  creator5: C5,
-  creator6: C6,
-  creator7: C7,
-  reducer: OnReducer<S, [C1, C2, C3, C4, C5, C6, C7]>
-): On<S>;
-export function on<
-  C1 extends ActionCreator,
-  C2 extends ActionCreator,
-  C3 extends ActionCreator,
-  C4 extends ActionCreator,
-  C5 extends ActionCreator,
-  C6 extends ActionCreator,
-  C7 extends ActionCreator,
-  C8 extends ActionCreator,
-  S
->(
-  creator1: C1,
-  creator2: C2,
-  creator3: C3,
-  creator4: C4,
-  creator5: C5,
-  creator6: C6,
-  creator7: C7,
-  creator8: C8,
-  reducer: OnReducer<S, [C1, C2, C3, C4, C5, C6, C7, C8]>
-): On<S>;
-export function on<
-  C1 extends ActionCreator,
-  C2 extends ActionCreator,
-  C3 extends ActionCreator,
-  C4 extends ActionCreator,
-  C5 extends ActionCreator,
-  C6 extends ActionCreator,
-  C7 extends ActionCreator,
-  C8 extends ActionCreator,
-  C9 extends ActionCreator,
-  S
->(
-  creator1: C1,
-  creator2: C2,
-  creator3: C3,
-  creator4: C4,
-  creator5: C5,
-  creator6: C6,
-  creator7: C7,
-  creator8: C8,
-  creator9: C9,
-  reducer: OnReducer<S, [C1, C2, C3, C4, C5, C6, C7, C8, C9]>
-): On<S>;
-export function on<
-  C1 extends ActionCreator,
-  C2 extends ActionCreator,
-  C3 extends ActionCreator,
-  C4 extends ActionCreator,
-  C5 extends ActionCreator,
-  C6 extends ActionCreator,
-  C7 extends ActionCreator,
-  C8 extends ActionCreator,
-  C9 extends ActionCreator,
-  C10 extends ActionCreator,
-  S
->(
-  creator1: C1,
-  creator2: C2,
-  creator3: C3,
-  creator4: C4,
-  creator5: C5,
-  creator6: C6,
-  creator7: C7,
-  creator8: C8,
-  creator9: C9,
-  creator10: C10,
-  reducer: OnReducer<S, [C1, C2, C3, C4, C5, C6, C7, C8, C9, C10]>
-): On<S>;
-export function on<S>(
-  creator: ActionCreator,
-  ...rest: (ActionCreator | OnReducer<S, [ActionCreator]>)[]
-): On<S>;
 /**
  * @description
  * Associates actions with a given state change function.
  * A state change function must be provided as the last parameter.
  *
  * @param args `ActionCreator`'s followed by a state change function.
  *
- * **To maintain type-safety**: pass 10 or less `ActionCreator`'s.
  * @returns an association of action types with a state change function.
+ *
+ * @example
+ * on(AuthApiActions.loginSuccess, (state, { user }) => ({ ...state, user }))
  */
-export function on(
-  ...args: (ActionCreator | Function)[]
-): { reducer: Function; types: string[] } {
-  const reducer = args.pop() as Function;
+export function on<
+  C extends ActionCreator[],
+  S,
+  OR extends OnReducer<S, C> = OnReducer<S, C>,
+  AR extends ActionReducer<S> = ActionReducer<S>
+>(...args: [...C, OR]): On<S> {
+  const reducer = (args.pop() as Function) as AR;
```

Comment:
```suggestion
export function on<
  Creators extends ActionCreator[],
  State,
  Reducer extends OnReducer<State, Creators>
>(...args: [...creators: Creators, reducer: Reducer]): On<State> {
  const reducer = (args.pop() as Function) as ActionReducer<State>;
```

Reply:
@timdeschryver
Btw, Prettier should be updated to [v2.1.0](https://prettier.io/blog/2020/08/24/2.1.0.html#support-for-typescript-40) or higher to support new TypeScript 4 features 🙂

Reply:
Good suggestion, thanks!
I'll update prettier in #2794